torch/_functorch/aot_autograd.py

```diff
@@ -1648,6 +1648,74 @@ def same_dtype_views(view1, view2):
     return True
 
 
+
+# Assumption: x and y are known to share a storage, and we are trying to determine
+# if their memory is actually completely disjoint, based on sizes/strides/storage_offset
+def tensors_definitely_do_not_overlap(x, y):
+    if x is y:
+        return False
+    if x.numel() == 0 or y.numel() == 0:
+        return True
+
+    # Make x always on the left
+    if x.storage_offset() > y.storage_offset():
+        x, y = y, x
+
+    # Short-circuit in the "obvious" overlapping case: both tensors are contiguous
+    if x.is_contiguous() and y.is_contiguous():
+        x_numbytes = x.numel() * x.itemsize
+        if x.storage_offset() + x_numbytes > y.storage_offset():
+            return False
+
+    # expected inputs: two lists of intervals in increasing order. Don't need to have the same length
+    def intervals_intersect(xs, ys):
+        if len(xs) == 0 or len(ys) == 0:
+            return False
+        x_ = xs[0]
+        y_ = ys[0]
+        if max(0, min(x_[1], y_[1]) - max(x_[0], y_[0])) > 0:
+            return True
+        if x_[1] > y_[1]:
+            return intervals_intersect(xs, ys[1:])
+        else:
+            return intervals_intersect(xs[1:], ys)
+
+    # This cases is needed for the shampoo optimizer.
```

Comment:
Not sure how shampoo helps here?
Example of two Tensors would be more useful here I think

Reply:
> Not sure how shampoo helps here?

Whoops thanks - the comment about shampoo was meant for lower down (the bit handling intervals)

> Example of two Tensors would be more useful here I think

You mean adding a comment that shows a small example of what this case looks like? Definitely, I can add one.

Reply:
Yes showing two small Tensors that would trigger this particular lack of overlap pattern.